src/components/views/Instance/Edit/InstancePermissions.tsx

```diff
@@ -0,0 +1,862 @@
+        });
+
+        const loadEnum = <
+            T extends Record<string, string | number>,
+            U extends keyof typeof newPermissionSet,
+            V extends keyof typeof this.state
+        >(
+            permEnum: T,
+            permSourceField: U,
+            permTargetField: V
+        ) => {
+            Object.entries(permEnum).forEach(([k, v]) => {
+                /* enums are objects that are reverse mapped, for example, an enum with a = 1 and b = 2 would look like this:
+                 * {
+                 *   "a": 1,
+                 *   "b": 2,
+                 *   1: "a",
+                 *   2: "b"
+                 * }
+                 * so we need to drop everything that doesnt resolve to a string because otherwise we end up with twice the values
+                 */
+                if (!isNaN(parseInt(k))) return;
+
+                const key = k.toLowerCase();
+                const val = v as number;
+
+                //we dont care about nothing
+                if (key == "none") return;
+
+                const currentVal = !!(newPermissionSet[permSourceField] & val);
+                this.setState(
+                    prevState =>
+                        (({
+                            [permTargetField]: {
+                                ...(prevState[permTargetField] as Record<string, Permission>),
+                                [key]: {
+                                    currentVal: currentVal,
+                                    bitflag: val
+                                }
+                            }
+                        } as unknown) as IState)
+                );
+            });
+        };
+
+        loadEnum(
+            InstancePermissionSetRights,
+            "instancePermissionSetRights",
+            "permsinstancepermissionset"
+        );
+        loadEnum(RepositoryRights, "repositoryRights", "permsrepository");
+        loadEnum(ByondRights, "byondRights", "permsbyond");
+        loadEnum(DreamMakerRights, "dreamMakerRights", "permsdreammaker");
+        loadEnum(DreamDaemonRights, "dreamDaemonRights", "permsdreamdaemon");
+        loadEnum(ChatBotRights, "chatBotRights", "permschatbots");
+        loadEnum(ConfigurationRights, "configurationRights", "permsconfiguration");
+    }
+
+    private addError(error: InternalError<ErrorCode>): void {
+        this.setState(prevState => {
+            const errors = Array.from(prevState.errors);
+            errors.push(error);
+            return {
+                errors
+            };
+        });
+    }
+
+    private async loadCurrentPermissions(selectedPermissionSetId: number) {
+        this.setState({
+            loadingPerms: true
+        });
+
+        if (selectedPermissionSetId === resolvePermissionSet(this.context.user).id) {
+            // This can hit the cache and seem "too" fast.
+            // slow it down a little
+            const sanicSpeedLimit = new Promise(r => setTimeout(r, 200));
```

Comment:
This seems a bit silly to me but if youre doing this, atleast initialize the promise where you are awaiting it, not 2 lines before

Reply:
But then you'd have to wait 200 on top of a _potential_ request